Mathlib/GroupTheory/DoubleCoset.lean

```diff
@@ -244,6 +253,94 @@ theorem right_bot_eq_right_quot (H : Subgroup G) :
   ext
   simp_rw [← rel_bot_eq_right_group_rel H]
 
+theorem iUnion_finset_quotToDoubleCoset (H K : Subgroup G) :
+    (∃ I : Finset (Quotient (H : Set G) K), ⋃ i ∈ I, quotToDoubleCoset H K i = .univ) ↔
+    Finite (Quotient (H : Set G) K) := by
+  constructor
+  · intro ⟨I, hI⟩
+    suffices (I : Set (Quotient (H : Set G) K)) = Set.univ by
+      rw [← Set.finite_univ_iff, ← this]
+      exact I.finite_toSet
+    rw [Set.eq_univ_iff_forall] at hI ⊢
+    rintro ⟨g⟩
+    obtain ⟨_, ⟨i, _, rfl⟩, T, ⟨hi, rfl⟩, hT : g ∈ quotToDoubleCoset H K i⟩ := hI g
+    rw [mem_quotToDoubleCoset_iff] at hT
+    simpa [← hT] using hi
+  · intro _
+    cases nonempty_fintype (Quotient (H : Set G) K)
+    use Finset.univ
+    simpa using union_quotToDoubleCoset H K
+
+theorem union_image_mk (H K : Subgroup G) :
+    ⋃ (q : Quotient H K), Quot.mk (QuotientGroup.leftRel K) ''
+    (doubleCoset (Quotient.out q : G) H K) = Set.univ ∧
+    ⋃ (q : Quotient H K), Quot.mk (QuotientGroup.rightRel H) ''
+    (doubleCoset (Quotient.out q : G) H K) = Set.univ := by
+  constructor
+  all_goals
+  have cover := union_quotToDoubleCoset H K
+  apply Set.Subset.antisymm_iff.mpr
+  simp only [Set.subset_univ, true_and]
+  intro x hx
+  simp only [Set.mem_iUnion, Set.mem_image]
+  obtain ⟨y, hy⟩ := Quot.exists_rep x
+  have ⟨i, hi⟩ : ∃ i : Quotient H K, y ∈ doubleCoset (Quotient.out i) H K  := by
+    contrapose cover
+    refine (Set.ne_univ_iff_exists_notMem _).mpr ?_
+    exact ⟨y, by simpa using cover⟩
+  exact ⟨i, y, hi, hy⟩
+
+theorem union_image_mk_leftRel (H K : Subgroup G) :
+    ⋃ (q : Quotient H K), Quot.mk (QuotientGroup.leftRel K) ''
+    (doubleCoset (Quotient.out q : G) H K) = Set.univ := (union_image_mk H K).1
+
+theorem union_image_mk_rightRel (H K : Subgroup G) :
+    ⋃ (q : Quotient H K), Quot.mk (QuotientGroup.rightRel H) ''
+    (doubleCoset (Quotient.out q : G) H K) = Set.univ := (union_image_mk H K).2
+
+theorem union_finset_leftRel_cover (H K : Subgroup G)
+    (t : Finset (Quotient H K)) (ht : Set.univ ⊆ ⋃ i ∈ t,
+    Quot.mk ⇑(QuotientGroup.leftRel K) '' doubleCoset (Quotient.out i)
+    H K) : ⋃ q ∈ t, doubleCoset (Quotient.out q) H K = Set.univ := by
```

Comment:
I think this would be more readable as:
```suggestion
theorem union_finset_leftRel_cover (H K : Subgroup G) (t : Finset (Quotient H K))
    (ht : Set.univ ⊆
      ⋃ i ∈ t, Quot.mk ⇑(QuotientGroup.leftRel K) '' doubleCoset (Quotient.out i) H K) :
    ⋃ q ∈ t, doubleCoset (Quotient.out q) H K = Set.univ := by
```